Installation failed: nvcc fatal

I failed to install ffpa-attn. It seems like occur a nvcc compilation error. 

Docker: nvcr.io/nvidia/pytorch:24.10-py3
Command: python3 setup.py bdist_wheel
Gpu Name: 4090
error message: 
```
--------------------------------------------------------------------FFPA-ATTN ENVs--------------------------------------------------------------------
PROJECT_DIR                   : /mnt/llm_dataset1/ezioliao/codes/ffpa-attn
ENABLE_FFPA_DEBUG             : False -> command: export ENABLE_FFPA_DEBUG=0
ENABLE_FFPA_ADA               : True  -> command: export ENABLE_FFPA_ADA=1
ENABLE_FFPA_AMPERE            : False -> command: export ENABLE_FFPA_AMPERE=0
ENABLE_FFPA_HOPPER            : False -> command: export ENABLE_FFPA_HOPPER=0
ENABLE_FFPA_ALL_STAGES        : True  -> command: export ENABLE_FFPA_ALL_STAGES=1
ENABLE_FFPA_ALL_HEADDIM       : False -> command: export ENABLE_FFPA_ALL_HEADDIM=0
ENABLE_FFPA_PREFETCH_QKV      : True  -> command: export ENABLE_FFPA_PREFETCH_QKV=1
ENABLE_FFPA_FORCE_QK_F16      : False -> command: export ENABLE_FFPA_FORCE_QK_F16=0
ENABLE_FFPA_FORCE_PV_F16      : False -> command: export ENABLE_FFPA_FORCE_PV_F16=0
ENABLE_FFPA_PERSIST_Q_G2S     : False -> command: export ENABLE_FFPA_PERSIST_Q_G2S=0
ENABLE_FFPA_PERSIST_KV_G2S    : False -> command: export ENABLE_FFPA_PERSIST_KV_G2S=0
ENABLE_FFPA_PERSIST_Q_S2R     : False -> command: export ENABLE_FFPA_PERSIST_Q_S2R=0
ENABLE_FFPA_PERSIST_V_S2R     : False -> command: export ENABLE_FFPA_PERSIST_V_S2R=0
ENABLE_FFPA_QKV_SMEM_SHARE    : False -> command: export ENABLE_FFPA_QKV_SMEM_SHARE=0
ENABLE_FFPA_SMEM_SWIZZLE_Q    : True  -> command: export ENABLE_FFPA_SMEM_SWIZZLE_Q=1
ENABLE_FFPA_SMEM_SWIZZLE_K    : True  -> command: export ENABLE_FFPA_SMEM_SWIZZLE_K=1
ENABLE_FFPA_SMEM_SWIZZLE_V    : True  -> command: export ENABLE_FFPA_SMEM_SWIZZLE_V=1
ENABLE_FFPA_REGISTERS_PIPE_KV : False -> command: export ENABLE_FFPA_REGISTERS_PIPE_KV=0
ENABLE_FFPA_LAUNCH_GRID_DNHB  : False -> command: export ENABLE_FFPA_LAUNCH_GRID_DNHB=0
------------------------------------------------------------------------------------------------------------------------------------------------------
------------------------------------------------------------------------------------------------------------------------------------------------------
csrc_file: /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/csrc/pybind/ffpa_attn_api.cc
csrc_file: /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/csrc/cuffpa/ffpa_attn_F16F16F16_L1.cu
csrc_file: /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/csrc/cuffpa/ffpa_attn_F16F16F32_L1.cu
------------------------------------------------------------------------------------------------------------------------------------------------------
running bdist_wheel
running build
running build_py
copying ffpa_attn/interface.py -> build/lib.linux-x86_64-cpython-310/ffpa_attn
copying ffpa_attn/__init__.py -> build/lib.linux-x86_64-cpython-310/ffpa_attn
copying ffpa_attn/version.py -> build/lib.linux-x86_64-cpython-310/ffpa_attn
running build_ext
building 'pyffpa_cuda' extension
creating /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/build/temp.linux-x86_64-cpython-310
creating /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/build/temp.linux-x86_64-cpython-310/mnt
creating /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/build/temp.linux-x86_64-cpython-310/mnt/llm_dataset1
creating /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/build/temp.linux-x86_64-cpython-310/mnt/llm_dataset1/ezioliao
creating /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/build/temp.linux-x86_64-cpython-310/mnt/llm_dataset1/ezioliao/codes
creating /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/build/temp.linux-x86_64-cpython-310/mnt/llm_dataset1/ezioliao/codes/ffpa-attn
creating /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/build/temp.linux-x86_64-cpython-310/mnt/llm_dataset1/ezioliao/codes/ffpa-attn/csrc
creating /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/build/temp.linux-x86_64-cpython-310/mnt/llm_dataset1/ezioliao/codes/ffpa-attn/csrc/cuffpa
creating /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/build/temp.linux-x86_64-cpython-310/mnt/llm_dataset1/ezioliao/codes/ffpa-attn/csrc/pybind
Emitting ninja build file /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/build/temp.linux-x86_64-cpython-310/build.ninja...
Compiling objects...
Allowing ninja to set a default number of workers... (overridable by setting the environment variable MAX_JOBS=N)
[1/3] /usr/local/cuda/bin/nvcc --generate-dependencies-with-compile --dependency-output /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/build/temp.linux-x86_64-cpython-310/mnt/llm_dataset1/ezioliao/codes/ffpa-attn/csrc/cuffpa/ffpa_attn_F16F16F16_L1.o.d -I/mnt/llm_dataset1/ezioliao/codes/ffpa-attn/include -I/usr/local/lib/python3.10/dist-packages/torch/include -I/usr/local/lib/python3.10/dist-packages/torch/include/torch/csrc/api/include -I/usr/local/lib/python3.10/dist-packages/torch/include/TH -I/usr/local/lib/python3.10/dist-packages/torch/include/THC -I/usr/local/cuda/include -I/usr/include/python3.10 -c -c /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/csrc/cuffpa/ffpa_attn_F16F16F16_L1.cu -o /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/build/temp.linux-x86_64-cpython-310/mnt/llm_dataset1/ezioliao/codes/ffpa-attn/csrc/cuffpa/ffpa_attn_F16F16F16_L1.o -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_BFLOAT16_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ --expt-relaxed-constexpr --compiler-options ''"'"'-fPIC'"'"'' -O3 -std=c++17 -U__CUDA_NO_HALF_OPERATORS__ -U__CUDA_NO_HALF_CONVERSIONS__ -U__CUDA_NO_HALF2_OPERATORS__ -U__CUDA_NO_BFLOAT16_CONVERSIONS__ --expt-relaxed-constexpr --expt-extended-lambda --use_fast_math --ptxas-options=-v --ptxas-options=-O3 '' -DBUILD_FFPA_ATTN_MMA_4090 '' -DENABLE_FFPA_ALL_STAGES -DENABLE_FFPA_PREFETCH_QKV -DENABLE_FFPA_SMEM_SWIZZLE_Q -DENABLE_FFPA_SMEM_SWIZZLE_K -DENABLE_FFPA_SMEM_SWIZZLE_V '-I /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/include' '-I /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/csrc/cuffpa' -gencode arch=compute_89,code=sm_89 -DTORCH_API_INCLUDE_EXTENSION_H '-DPYBIND11_COMPILER_TYPE="_gcc"' '-DPYBIND11_STDLIB="_libstdcpp"' '-DPYBIND11_BUILD_ABI="_cxxabi1016"' -DTORCH_EXTENSION_NAME=pyffpa_cuda -D_GLIBCXX_USE_CXX11_ABI=1
FAILED: /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/build/temp.linux-x86_64-cpython-310/mnt/llm_dataset1/ezioliao/codes/ffpa-attn/csrc/cuffpa/ffpa_attn_F16F16F16_L1.o 
/usr/local/cuda/bin/nvcc --generate-dependencies-with-compile --dependency-output /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/build/temp.linux-x86_64-cpython-310/mnt/llm_dataset1/ezioliao/codes/ffpa-attn/csrc/cuffpa/ffpa_attn_F16F16F16_L1.o.d -I/mnt/llm_dataset1/ezioliao/codes/ffpa-attn/include -I/usr/local/lib/python3.10/dist-packages/torch/include -I/usr/local/lib/python3.10/dist-packages/torch/include/torch/csrc/api/include -I/usr/local/lib/python3.10/dist-packages/torch/include/TH -I/usr/local/lib/python3.10/dist-packages/torch/include/THC -I/usr/local/cuda/include -I/usr/include/python3.10 -c -c /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/csrc/cuffpa/ffpa_attn_F16F16F16_L1.cu -o /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/build/temp.linux-x86_64-cpython-310/mnt/llm_dataset1/ezioliao/codes/ffpa-attn/csrc/cuffpa/ffpa_attn_F16F16F16_L1.o -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_BFLOAT16_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ --expt-relaxed-constexpr --compiler-options ''"'"'-fPIC'"'"'' -O3 -std=c++17 -U__CUDA_NO_HALF_OPERATORS__ -U__CUDA_NO_HALF_CONVERSIONS__ -U__CUDA_NO_HALF2_OPERATORS__ -U__CUDA_NO_BFLOAT16_CONVERSIONS__ --expt-relaxed-constexpr --expt-extended-lambda --use_fast_math --ptxas-options=-v --ptxas-options=-O3 '' -DBUILD_FFPA_ATTN_MMA_4090 '' -DENABLE_FFPA_ALL_STAGES -DENABLE_FFPA_PREFETCH_QKV -DENABLE_FFPA_SMEM_SWIZZLE_Q -DENABLE_FFPA_SMEM_SWIZZLE_K -DENABLE_FFPA_SMEM_SWIZZLE_V '-I /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/include' '-I /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/csrc/cuffpa' -gencode arch=compute_89,code=sm_89 -DTORCH_API_INCLUDE_EXTENSION_H '-DPYBIND11_COMPILER_TYPE="_gcc"' '-DPYBIND11_STDLIB="_libstdcpp"' '-DPYBIND11_BUILD_ABI="_cxxabi1016"' -DTORCH_EXTENSION_NAME=pyffpa_cuda -D_GLIBCXX_USE_CXX11_ABI=1
nvcc fatal   : A single input file is required for a non-link phase when an outputfile is specified
[2/3] /usr/local/cuda/bin/nvcc --generate-dependencies-with-compile --dependency-output /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/build/temp.linux-x86_64-cpython-310/mnt/llm_dataset1/ezioliao/codes/ffpa-attn/csrc/cuffpa/ffpa_attn_F16F16F32_L1.o.d -I/mnt/llm_dataset1/ezioliao/codes/ffpa-attn/include -I/usr/local/lib/python3.10/dist-packages/torch/include -I/usr/local/lib/python3.10/dist-packages/torch/include/torch/csrc/api/include -I/usr/local/lib/python3.10/dist-packages/torch/include/TH -I/usr/local/lib/python3.10/dist-packages/torch/include/THC -I/usr/local/cuda/include -I/usr/include/python3.10 -c -c /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/csrc/cuffpa/ffpa_attn_F16F16F32_L1.cu -o /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/build/temp.linux-x86_64-cpython-310/mnt/llm_dataset1/ezioliao/codes/ffpa-attn/csrc/cuffpa/ffpa_attn_F16F16F32_L1.o -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_BFLOAT16_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ --expt-relaxed-constexpr --compiler-options ''"'"'-fPIC'"'"'' -O3 -std=c++17 -U__CUDA_NO_HALF_OPERATORS__ -U__CUDA_NO_HALF_CONVERSIONS__ -U__CUDA_NO_HALF2_OPERATORS__ -U__CUDA_NO_BFLOAT16_CONVERSIONS__ --expt-relaxed-constexpr --expt-extended-lambda --use_fast_math --ptxas-options=-v --ptxas-options=-O3 '' -DBUILD_FFPA_ATTN_MMA_4090 '' -DENABLE_FFPA_ALL_STAGES -DENABLE_FFPA_PREFETCH_QKV -DENABLE_FFPA_SMEM_SWIZZLE_Q -DENABLE_FFPA_SMEM_SWIZZLE_K -DENABLE_FFPA_SMEM_SWIZZLE_V '-I /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/include' '-I /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/csrc/cuffpa' -gencode arch=compute_89,code=sm_89 -DTORCH_API_INCLUDE_EXTENSION_H '-DPYBIND11_COMPILER_TYPE="_gcc"' '-DPYBIND11_STDLIB="_libstdcpp"' '-DPYBIND11_BUILD_ABI="_cxxabi1016"' -DTORCH_EXTENSION_NAME=pyffpa_cuda -D_GLIBCXX_USE_CXX11_ABI=1
FAILED: /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/build/temp.linux-x86_64-cpython-310/mnt/llm_dataset1/ezioliao/codes/ffpa-attn/csrc/cuffpa/ffpa_attn_F16F16F32_L1.o 
/usr/local/cuda/bin/nvcc --generate-dependencies-with-compile --dependency-output /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/build/temp.linux-x86_64-cpython-310/mnt/llm_dataset1/ezioliao/codes/ffpa-attn/csrc/cuffpa/ffpa_attn_F16F16F32_L1.o.d -I/mnt/llm_dataset1/ezioliao/codes/ffpa-attn/include -I/usr/local/lib/python3.10/dist-packages/torch/include -I/usr/local/lib/python3.10/dist-packages/torch/include/torch/csrc/api/include -I/usr/local/lib/python3.10/dist-packages/torch/include/TH -I/usr/local/lib/python3.10/dist-packages/torch/include/THC -I/usr/local/cuda/include -I/usr/include/python3.10 -c -c /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/csrc/cuffpa/ffpa_attn_F16F16F32_L1.cu -o /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/build/temp.linux-x86_64-cpython-310/mnt/llm_dataset1/ezioliao/codes/ffpa-attn/csrc/cuffpa/ffpa_attn_F16F16F32_L1.o -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_BFLOAT16_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ --expt-relaxed-constexpr --compiler-options ''"'"'-fPIC'"'"'' -O3 -std=c++17 -U__CUDA_NO_HALF_OPERATORS__ -U__CUDA_NO_HALF_CONVERSIONS__ -U__CUDA_NO_HALF2_OPERATORS__ -U__CUDA_NO_BFLOAT16_CONVERSIONS__ --expt-relaxed-constexpr --expt-extended-lambda --use_fast_math --ptxas-options=-v --ptxas-options=-O3 '' -DBUILD_FFPA_ATTN_MMA_4090 '' -DENABLE_FFPA_ALL_STAGES -DENABLE_FFPA_PREFETCH_QKV -DENABLE_FFPA_SMEM_SWIZZLE_Q -DENABLE_FFPA_SMEM_SWIZZLE_K -DENABLE_FFPA_SMEM_SWIZZLE_V '-I /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/include' '-I /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/csrc/cuffpa' -gencode arch=compute_89,code=sm_89 -DTORCH_API_INCLUDE_EXTENSION_H '-DPYBIND11_COMPILER_TYPE="_gcc"' '-DPYBIND11_STDLIB="_libstdcpp"' '-DPYBIND11_BUILD_ABI="_cxxabi1016"' -DTORCH_EXTENSION_NAME=pyffpa_cuda -D_GLIBCXX_USE_CXX11_ABI=1
nvcc fatal   : A single input file is required for a non-link phase when an outputfile is specified
[3/3] c++ -MMD -MF /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/build/temp.linux-x86_64-cpython-310/mnt/llm_dataset1/ezioliao/codes/ffpa-attn/csrc/pybind/ffpa_attn_api.o.d -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -I/mnt/llm_dataset1/ezioliao/codes/ffpa-attn/include -I/usr/local/lib/python3.10/dist-packages/torch/include -I/usr/local/lib/python3.10/dist-packages/torch/include/torch/csrc/api/include -I/usr/local/lib/python3.10/dist-packages/torch/include/TH -I/usr/local/lib/python3.10/dist-packages/torch/include/THC -I/usr/local/cuda/include -I/usr/include/python3.10 -c -c /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/csrc/pybind/ffpa_attn_api.cc -o /mnt/llm_dataset1/ezioliao/codes/ffpa-attn/build/temp.linux-x86_64-cpython-310/mnt/llm_dataset1/ezioliao/codes/ffpa-attn/csrc/pybind/ffpa_attn_api.o -O3 -std=c++17 -DTORCH_API_INCLUDE_EXTENSION_H '-DPYBIND11_COMPILER_TYPE="_gcc"' '-DPYBIND11_STDLIB="_libstdcpp"' '-DPYBIND11_BUILD_ABI="_cxxabi1016"' -DTORCH_EXTENSION_NAME=pyffpa_cuda -D_GLIBCXX_USE_CXX11_ABI=1
ninja: build stopped: subcommand failed.
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/cpp_extension.py", line 2104, in _run_ninja_build
    subprocess.run(
  File "/usr/lib/python3.10/subprocess.py", line 526, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['ninja', '-v']' returned non-zero exit status 1.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/mnt/llm_dataset1/ezioliao/codes/ffpa-attn/setup.py", line 70, in <module>
    setup(
  File "/usr/local/lib/python3.10/dist-packages/setuptools/__init__.py", line 103, in setup
    return distutils.core.setup(**attrs)
  File "/usr/local/lib/python3.10/dist-packages/setuptools/_distutils/core.py", line 184, in setup
    return run_commands(dist)
  File "/usr/local/lib/python3.10/dist-packages/setuptools/_distutils/core.py", line 200, in run_commands
    dist.run_commands()
  File "/usr/local/lib/python3.10/dist-packages/setuptools/_distutils/dist.py", line 970, in run_commands
    self.run_command(cmd)
  File "/usr/local/lib/python3.10/dist-packages/setuptools/dist.py", line 974, in run_command
    super().run_command(command)
  File "/usr/local/lib/python3.10/dist-packages/setuptools/_distutils/dist.py", line 989, in run_command
    cmd_obj.run()
  File "/usr/local/lib/python3.10/dist-packages/setuptools/command/bdist_wheel.py", line 373, in run
    self.run_command("build")
  File "/usr/local/lib/python3.10/dist-packages/setuptools/_distutils/cmd.py", line 316, in run_command
    self.distribution.run_command(command)
  File "/usr/local/lib/python3.10/dist-packages/setuptools/dist.py", line 974, in run_command
    super().run_command(command)
  File "/usr/local/lib/python3.10/dist-packages/setuptools/_distutils/dist.py", line 989, in run_command
    cmd_obj.run()
  File "/usr/local/lib/python3.10/dist-packages/setuptools/_distutils/command/build.py", line 135, in run
    self.run_command(cmd_name)
  File "/usr/local/lib/python3.10/dist-packages/setuptools/_distutils/cmd.py", line 316, in run_command
    self.distribution.run_command(command)
  File "/usr/local/lib/python3.10/dist-packages/setuptools/dist.py", line 974, in run_command
    super().run_command(command)
  File "/usr/local/lib/python3.10/dist-packages/setuptools/_distutils/dist.py", line 989, in run_command
    cmd_obj.run()
  File "/usr/local/lib/python3.10/dist-packages/setuptools/command/build_ext.py", line 93, in run
    _build_ext.run(self)
  File "/usr/local/lib/python3.10/dist-packages/setuptools/_distutils/command/build_ext.py", line 359, in run
    self.build_extensions()
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/cpp_extension.py", line 868, in build_extensions
    build_ext.build_extensions(self)
  File "/usr/local/lib/python3.10/dist-packages/setuptools/_distutils/command/build_ext.py", line 479, in build_extensions
    self._build_extensions_serial()
  File "/usr/local/lib/python3.10/dist-packages/setuptools/_distutils/command/build_ext.py", line 505, in _build_extensions_serial
    self.build_extension(ext)
  File "/usr/local/lib/python3.10/dist-packages/setuptools/command/build_ext.py", line 254, in build_extension
    _build_ext.build_extension(self, ext)
  File "/usr/local/lib/python3.10/dist-packages/Cython/Distutils/build_ext.py", line 135, in build_extension
    super(build_ext, self).build_extension(ext)
  File "/usr/local/lib/python3.10/dist-packages/setuptools/_distutils/command/build_ext.py", line 560, in build_extension
    objects = self.compiler.compile(
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/cpp_extension.py", line 681, in unix_wrap_ninja_compile
    _write_ninja_file_and_compile_objects(
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/cpp_extension.py", line 1784, in _write_ninja_file_and_compile_objects
    _run_ninja_build(
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/cpp_extension.py", line 2120, in _run_ninja_build
    raise RuntimeError(message) from e
RuntimeError: Error compiling objects for extension
```

--- a/env.py
+++ b/env.py
@@ -338,17 +338,18 @@ class ENV(object):
         extra_cuda_cflags.append("--expt-relaxed-constexpr")
         extra_cuda_cflags.append("--expt-extended-lambda")
         extra_cuda_cflags.append("--use_fast_math")
+        extra_cuda_cflags.append(_specific_device_macro())
+        extra_cuda_cflags.extend(ENV.env_cuda_cflags())
+        extra_cuda_cflags.append(f"-I {ENV.project_dir()}/include")
+        extra_cuda_cflags.append(f"-I {ENV.project_dir()}/csrc/cuffpa")
         extra_cuda_cflags.append(
             "-diag-suppress 177" if not build_pkg else "--ptxas-options=-v"
         )
         extra_cuda_cflags.append(
             "-Xptxas -v" if not build_pkg else "--ptxas-options=-O3"
         )
-        if _specific_device_macro() is not None:
-            extra_cuda_cflags.append(_specific_device_macro())
-        extra_cuda_cflags.extend(ENV.env_cuda_cflags())
-        extra_cuda_cflags.append(f"-I {ENV.project_dir()}/include")
-        extra_cuda_cflags.append(f"-I {ENV.project_dir()}/csrc/cuffpa")
+        # Avoid None or empty str as flag or macro
+        extra_cuda_cflags = [flag for flag in extra_cuda_cflags if flag]
         return extra_cuda_cflags
 
     @staticmethod

--- a/env.py
+++ b/env.py
@@ -178,7 +178,7 @@ class ENV(object):
         if cls.enable_persist_kv_g2s():
             return cls.ENABLE_FFPA_PERSIST_V_S2R
         return False
-    
+
     @classmethod
     def enable_registers_pipe_kv(cls):
         return cls.ENABLE_FFPA_REGISTERS_PIPE_KV
@@ -312,6 +312,22 @@ class ENV(object):
     @staticmethod
     def get_build_cuda_cflags(build_pkg: bool = False):
         device_name = ENV.get_device_name()
+
+        def _specific_device_tag():
+            if "L20" in device_name:
+                return "L20"
+            elif "4090" in device_name:
+                return "4090"
+            elif "3080" in device_name:
+                return "3080"
+            return None
+
+        def _specific_device_macro():
+            tag = _specific_device_tag()
+            if tag is not None:
+                return f"-DBUILD_FFPA_ATTN_MMA_{tag}"
+            return None
+
         extra_cuda_cflags = []
         extra_cuda_cflags.append("-O3")
         extra_cuda_cflags.append("-std=c++17")
@@ -328,15 +344,8 @@ class ENV(object):
         extra_cuda_cflags.append(
             "-Xptxas -v" if not build_pkg else "--ptxas-options=-O3"
         )
-        extra_cuda_cflags.append(
-            "-DBUILD_FFPA_ATTN_MMA_L20" if "L20" in device_name else ""
-        )
-        extra_cuda_cflags.append(
-            "-DBUILD_FFPA_ATTN_MMA_4090" if "4090" in device_name else ""
-        )
-        extra_cuda_cflags.append(
-            "-DBUILD_FFPA_ATTN_MMA_3080" if "3080" in device_name else ""
-        )
+        if _specific_device_macro() is not None:
+            extra_cuda_cflags.append(_specific_device_macro())
         extra_cuda_cflags.extend(ENV.env_cuda_cflags())
         extra_cuda_cflags.append(f"-I {ENV.project_dir()}/include")
         extra_cuda_cflags.append(f"-I {ENV.project_dir()}/csrc/cuffpa")

